lib/std/mem.zig

```diff
@@ -279,15 +279,62 @@ pub fn set(comptime T: type, dest: []T, value: T) void {
 /// Zero initializes the type.
 /// This can be used to zero initialize a C-struct.
 pub fn zeroes(comptime T: type) T {
-    if (@sizeOf(T) == 0) return T{};
-
-    if (comptime meta.containerLayout(T) != .Extern) {
-        @compileError("TODO: Currently this only works for extern types");
+    switch (@typeInfo(T)) {
+        .ComptimeInt, .Int, .ComptimeFloat, .Float => {
+            return @as(T, 0);
+        },
+        .Enum, .EnumLiteral => {
+            return @intToEnum(T, 0);
+        },
+        .Void => {
+            return {};
+        },
+        .Bool => {
+            return false;
+        },
+        .Optional, .Null => {
+            return null;
+        },
+        .Struct => {
+            if (@sizeOf(T) == 0) return T{};
+            if (comptime meta.containerLayout(T) == .Extern) {
+                var item: T = undefined;
+                @memset(@ptrCast([*]u8, &item), 0, @sizeOf(T));
+                return item;
+            } else {
+                var structure: T = undefined;
+                comptime var field_i = 0;
+                inline while (field_i < @memberCount(T)) : (field_i += 1) {
+                    @field(structure, @memberName(T, field_i)) = zeroes(@TypeOf(@field(structure, @memberName(T, field_i))));
+                }
+                return structure;
+            }
+        },
+        .Pointer => |ptr_info| {
+            if (ptr_info.is_allowzero) {
+                return null;
```

Comment:
I think if you add test coverage for this case you'll find a compile error. 

Probably the best bet would be to ignore the "allowzero" attribute, and handle it with the `@compileError` below like any other pointer. An `allowzero` pointer means that the address 0 may have valid data there - which means that initializing it to 0 is equivalent to initializing a normal pointer to an arbitrary address.

Reply:
Ok, makes sense.